### DISCUSSION TITLE
socketio.exceptions.BadNamespaceError: / is not a connected namespace.

### DISCUSSION THREAD

**Discussion by mishkat-ucr** [miguelgrinberg/Flask-SocketIO#1589] [Q&A]
I am trying out one use case here, which is: send over some data in batches to the client. I will initiate an event from the client, in which server will respond with another event which will send the next batch. In a nutshell, I'm trying to achieve some sort of synchronicity in an asynchronous event based mechanism.

Here is the gist of my client code:
```
@socketio_client.on('continue train')
def continue_train():
        socketio_client.emit('train', 'hello world', callback=ack_train)


@socketio_client.on('continue test')
def continue_test():
        socketio_client.emit('test', 'hello world', callback=ack_test)

socketio_client.emit('train', 'hello world', callback=ack_train)

socketio_client.emit('train', 'hello world', callback=ack_train)
``` 

Here is the gist of my server code:
```
@socketio.on('train')
def handle_train(data):
    print(data)
    emit('continue train')


@socketio.on('test')
def handle_test(data):
    print(data)
    emit('continue test')
```

However, I am getting a `socketio.exceptions.BadNamespaceError: / is not a connected namespace.` error. Any idea why?

**Comment by miguelgrinberg** [comment]
Yes, this issue is known and is explained here: https://github.com/miguelgrinberg/python-socketio/issues/634.

The master branch of python-socketio has a fix, but it hasn't been released yet. Feel free to install from git if you want to try out the fix, or else you can also change your application as suggested in the issue linked.

**Comment by mishkat-ucr** [comment]
Great, @miguelgrinberg ! Thanks.

**Comment by mishkat-ucr** [comment]
Hi @miguelgrinberg , I am using the 5.0.2.dev0 version now. But, the problem still persists. I am using a sio.sleep(1) after the connect as well. Any insights after this?

**Comment by miguelgrinberg** [comment]
@mishkat-ucr the fix is not in this repo, it's in python-socketio. That's the one you need to install from git.

**Comment by younengma** [comment]
face connecting error 
while try connecting to server. 

1. if i use the latest version, it will report " socketio.exceptions.ConnectionError: One or more namespaces failed to connect"
python version 
flask_socketio==5.1.0  (also tried 5.1.1)
python-socketio==5.3.0  (also tried 5.4.0)
python-engineio==4.2.1 (also tried 4.2.0)

I am able to connect to the server using java socket client with version:
 <groupId>io.socket</groupId>
 <artifactId>socket.io-client</artifactId>
 <version>2.0.1</version>

2. if i change the python-socketio==5.0.2, I will  report  ”socketio.exceptions.BadNamespaceError: / is not a connected namespace.“

3. if  install  flsk_socketio==4.3.2,  the python-socketio and engineio will be lowered and the python client can connect succesully. 

It seems that there is some issue on the python socket client side on the lastest version 5.3.x above.

**Reply by miguelgrinberg** [reply]
See the version compatibility section of the documentation. The problem here is that your server is using an old version of the Socket.IO protocol.
